pkg/ingestor/parser/spdx/parse_spdx.go

```diff
@@ -253,13 +251,20 @@ func (s *spdxParser) GetPredicates(ctx context.Context) *assembler.IngestPredica
 			logger.Errorf("SPDX document had invalid created time %q : %w", s.spdxDoc.CreationInfo.Created, err)
 			return nil
 		}
-		for _, topLevelPkg := range topLevel {
-			preds.HasSBOM = append(preds.HasSBOM, common.CreateTopLevelHasSBOM(topLevelPkg, s.doc, s.spdxDoc.DocumentNamespace, timestamp))
+
+		if len(topLevelArts) > 0 {
+			for _, topLevelArt := range topLevelArts {
+				preds.HasSBOM = append(preds.HasSBOM, common.CreateTopLevelHasSBOMFromArtifact(topLevelArt, s.doc, s.spdxDoc.DocumentNamespace, timestamp))
+			}
+		} else {
+			for _, topLevelPkg := range topLevelPkgs {
+				preds.HasSBOM = append(preds.HasSBOM, common.CreateTopLevelHasSBOMFromPkg(topLevelPkg, s.doc, s.spdxDoc.DocumentNamespace, timestamp))
+			}
```

Comment:
Hmm, I am not quite clear on what you are talking about here. Can you clarify, please?

Perhaps you mean an SPDX JSON SBOM like the following?

```json
{
  "spdxVersion": "SPDX-2.3",
  "dataLicense": "CC0-1.0",
  "SPDXID": "SPDXRef-DOCUMENT",
  "creationInfo": {
    "created": "2024-04-30T01:12:27Z"
  },
  "name": "for-testing-with-checksum",
  "documentNamespace": "https://example.com/for-testing-with-checksum",
  "packages": [
    {
      "name": "for-testing-with-checksum-pkg",
      "SPDXID": "SPDXRef-Package-for-testing-with-checksum-pkg",
      "downloadLocation": "https://example.com/for-testing-with-checksum-pkg",
      "checksums": [
        {
          "algorithm": "SHA1",
          "checksumValue": "pkgsha1"
        },
        {
          "algorithm": "SHA3-384",
          "checksumValue": "pkgsha3-384"
        }
      ]
    },
    {
      "name": "for-testing-without-checksum-pkg",
      "SPDXID": "SPDXRef-Package-for-testing-without-checksum-pkg",
      "downloadLocation": "https://example.com/for-testing-without-checksum-pkg"
    }
  ],
  "relationships": [
    {
      "spdxElementId": "SPDXRef-DOCUMENT",
      "relationshipType": "DESCRIBES",
      "relatedSpdxElement": "SPDXRef-Package-for-testing-with-checksum-pkg"
    },
    {
      "spdxElementId": "SPDXRef-Package-for-testing-with-checksum-pkg",
      "relationshipType": "DESCRIBED_BY",
      "relatedSpdxElement": "SPDXRef-DOCUMENT"
    },
    {
      "spdxElementId": "SPDXRef-Package-for-testing-without-checksum-pkg",
      "relationshipType": "DESCRIBED_BY",
      "relatedSpdxElement": "SPDXRef-DOCUMENT"
    }
  ]
}
```

What I mean here is that: there is an extra `relationships` value, a `DESCRIBED_BY` that lists another package, and this package does not have a checksum. (The way I came to this was: 1. Q: How do we create `HasSBOM`s? A: From top-level artifacts or top-level packages. 2. How do we come up with top-level artifacts and packages? A: By looking at packages and artifacts whose SPDX ID is a top-level one. 3. Q: How we know which SPDX IDs are top-level ones? A: In `s.getTopLevelSPDXIDs()`, we identify such IDs using the `relationships` array in the SBOM JSON, and from previous conversations with Parth, an SPDX SBOM doc can `DESCRIBE` only thing, so I think you either mean a package or artifact `DESCRIBED_BY` the doc, and if it was an artifact then we would have a hash/digest/checksum value, so you must mean a package. Either this chain of reasoning is correct, or I am misunderstanding.)

If this is what you mean, then I can try to come up with a way to identify such cases (using the example I provided as a test-case) and make sure that we log when such an event occurs. I do note that now, we will create `hasSBOM` nodes for the package artifact, but the checksum-less package is ignored. I think that this is what you mean.

Reply:
Yup, that's what I meant! If there is a package with a checksum and a package without, and both are top level packages, then an SBOM for the one without the checksum will not be created. I think this could be checked simply by comparing the lengths of `topLevelPackages` and `topLevelArtifacts`.

Reply:
I found that your idea of comparing the lengths of the top-level pkgs and top-level artifacts slices was almost correct but there can be multiple artifacts for a single pkg and this throws things off. So what I did instead was to store top-level artifacts in a map keyed off package SPDXIDs whose values are all the artifacts for that package SPDXID. Comparing the lengths of this "(unique) top-level packages with artifacts" map with the length of the top-level packages seems to work. Fixed in 8fe017d88422d9ecfe25d27ef0dd0ebcff6af426.